app/src/main/java/org/mozilla/fenix/browser/ToolbarGestureHandler.kt

```diff
@@ -0,0 +1,374 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.browser
+
+import android.animation.Animator
+import android.animation.AnimatorListenerAdapter
+import android.app.Activity
+import android.graphics.PointF
+import android.graphics.Rect
+import android.os.Build
+import android.util.DisplayMetrics
+import android.util.TypedValue
+import android.view.View
+import android.view.ViewConfiguration
+import androidx.core.graphics.contains
+import androidx.core.graphics.toPoint
+import androidx.core.view.WindowInsetsCompat
+import androidx.core.view.doOnAttach
+import androidx.core.view.isVisible
+import androidx.dynamicanimation.animation.DynamicAnimation
+import androidx.dynamicanimation.animation.FlingAnimation
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.Main
+import kotlinx.coroutines.launch
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.support.ktx.android.util.dpToPx
+import mozilla.components.support.ktx.android.view.getRectWithViewLocation
+import org.mozilla.fenix.ext.sessionsOfType
+import org.mozilla.fenix.ext.settings
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.min
+
+/**
+ * Handles intercepting touch events on the toolbar for swipe gestures and executes the
+ * necessary animations.
+ */
+@Suppress("LargeClass", "TooManyFunctions")
+class ToolbarGestureHandler(
+    private val activity: Activity,
+    private val contentLayout: View,
+    private val tabPreview: TabPreview,
+    private val toolbarLayout: View,
+    private val sessionManager: SessionManager
+) : SwipeGestureListener {
+
+    private enum class GestureDirection {
+        LEFT_TO_RIGHT, RIGHT_TO_LEFT
+    }
+
+    private sealed class Destination {
+        data class Tab(val session: Session) : Destination()
+        object None : Destination()
+    }
+
+    private var windowInsets: WindowInsetsCompat? = null
+
+    private val windowWidth: Int
+        get() = with(DisplayMetrics()) {
+            activity.windowManager.defaultDisplay.getMetrics(this)
+            this.widthPixels
+        }
+
+    private val previewOffset = with(DisplayMetrics()) {
+        activity.windowManager.defaultDisplay.getMetrics(this)
+        PREVIEW_OFFSET_DP.dpToPx(this)
+    }
+
+    private val touchSlop = ViewConfiguration.get(activity).scaledTouchSlop
+    private val minimumFlingVelocity = ViewConfiguration.get(activity).scaledMinimumFlingVelocity
+    private val defaultVelocity = TypedValue.applyDimension(
+        TypedValue.COMPLEX_UNIT_DIP,
+        DP_PER_SECOND,
+        activity.resources.displayMetrics
+    )
+
+    private var gestureDirection = GestureDirection.LEFT_TO_RIGHT
+
+    init {
+        CoroutineScope(Main).launch {
```

Comment:
Are we launching here for a race condition? Is there another way we can fix it? (If not, we should probably at least attach it to a lifecycleScope)

Reply:
Went with using a lifecycleScope. In theory the rootview should already be attached so we can get the insets right away, but I don't want to risk that not always being the case and crashing.

Reply:
I found one scenario where the insets can change (switching from gesture -> three-button nav or vice-versa) so I decided to get the insets everytime I need to check them. To do that I changed `getInsets` so it'll just return null if the `decorView` isn't attached. In theory, we shouldn't run into this scenario and we'll be able to get the insets.